modules/git/commit_info.go

```diff
@@ -5,325 +5,233 @@
+	}
 
-	/* mutable fields, must hold mutex to read or write */
-	// map from filepath to commit
-	commits map[string]*Commit
-	// set of filepaths that have been or are being searched for in a target search
-	targetedPaths map[string]struct{}
-}
+	revs, err := getLastCommitForPaths(c, treePath, entryPaths)
+	if err != nil {
+		return nil, nil, err
+	}
 
-func (state *getCommitsInfoState) numRemainingEntries() int {
-	state.lock.Lock()
-	defer state.lock.Unlock()
-	return len(state.entries) - len(state.commits)
-}
+	commit.repo.gogitStorage.Close()
 
-// getTargetEntryPath Returns the next path for a targeted-searching thread to
-// search for, or returns the empty string if nothing left to search for
-func (state *getCommitsInfoState) getTargetedEntryPath() string {
-	var targetedEntryPath string
-	state.lock.Lock()
-	defer state.lock.Unlock()
-	for _, entry := range state.entries {
-		entryPath := path.Join(state.treePath, entry.Name())
-		if _, ok := state.commits[entryPath]; ok {
-			continue
-		} else if _, ok = state.targetedPaths[entryPath]; ok {
-			continue
+	commitsInfo := make([][]interface{}, len(tes))
+	for i, entry := range tes {
+		if rev, ok := revs[entry.Name()]; ok {
+			entryCommit := convertCommit(rev)
+			if entry.IsSubModule() {
+				subModuleURL := ""
+				if subModule, err := commit.GetSubModule(entry.Name()); err != nil {
+					return nil, nil, err
+				} else if subModule != nil {
+					subModuleURL = subModule.URL
+				}
+				subModuleFile := NewSubModuleFile(entryCommit, subModuleURL, entry.ID.String())
+				commitsInfo[i] = []interface{}{entry, subModuleFile}
+			} else {
+				commitsInfo[i] = []interface{}{entry, entryCommit}
+			}
+		} else {
+			commitsInfo[i] = []interface{}{entry, nil}
 		}
-		targetedEntryPath = entryPath
-		state.targetedPaths[entryPath] = struct{}{}
-		break
 	}
-	return targetedEntryPath
-}
 
-// repeatedly perform targeted searches for unpopulated entries
-func targetedSearch(state *getCommitsInfoState, done chan error, cache LastCommitCache) {
-	for {
-		entryPath := state.getTargetedEntryPath()
-		if len(entryPath) == 0 {
-			done <- nil
-			return
-		}
-		if cache != nil {
-			commit, err := cache.Get(state.headCommit.repo.Path, state.headCommit.ID.String(), entryPath)
-			if err == nil && commit != nil {
-				state.update(entryPath, commit)
-				continue
-			}
-		}
-		command := NewCommand("rev-list", "-1", state.headCommit.ID.String(), "--", entryPath)
-		output, err := command.RunInDir(state.headCommit.repo.Path)
-		if err != nil {
-			done <- err
-			return
-		}
-		id, err := NewIDFromString(strings.TrimSpace(output))
-		if err != nil {
-			done <- err
-			return
-		}
-		commit, err := state.headCommit.repo.getCommit(id)
-		if err != nil {
-			done <- err
-			return
-		}
-		state.update(entryPath, commit)
-		if cache != nil {
-			cache.Put(state.headCommit.repo.Path, state.headCommit.ID.String(), entryPath, commit)
-		}
+	var treeCommit *Commit
+	if rev, ok := revs[""]; ok {
```

Comment:
Can we add some comments here why `revs[""]` maybe exist?

Reply:
Sure, I will add a commit. It's an optimization that helps with the listing page. Gitea displays the most recent commit from the actual directory on the top and it's efficient to get it alongside the commits for individual files.